### PR TITLE
When checking for zero target rates, also check wells under group control where required

### PR DESCRIPTION
**Pull request by steink**
Several places in the code we check whether a well is set to inject/produce at zero rate since this case requires special treatment. This is currently done by the function `wellUnderZeroRateTarget` (often called from `stopppedOrZeroRateTarget`). Wells under group control are currently not included in this check, although the _special treatments_ are equally required if a well is assigned zero rate from a group.

I've written draft versions of the two above mentioned functions that also take group-controlled wells into account. Would just like some feedback/discussion whether this seems reasonable before I go ahead and suggest updating all relevant function calls (the new functions have temporary names `*Version`).   

I've verified that this approach fixes an issue with `iterateWellEqWithSwitching` which currently does not behave nicely for wells that get assigned zero rate from group.

### CODE REVIEW

**Comment by GitPaean** [comment]
jenkins build this please

File: opm/simulators/wells/WellInterface_impl.hpp
Comment (GitPaean) [inline opm/simulators/wells/WellInterface_impl.hpp:1402]: I think you can use `wellUnderZeroRateTargetVersion` to replace the original `wellUnderZeroRateTarget`

File: opm/simulators/wells/WellInterface_impl.hpp
Comment (GitPaean) [inline opm/simulators/wells/WellInterface_impl.hpp:1416]: I think the only reason we need `Simulator` is because of `schedule`, maybe also the `GroupState`. We can pass in the `schedule` as a parameter, so the function can be moved to `WellInterfaceGeneric`.

File: opm/simulators/wells/WellInterface_impl.hpp
Comment (GitPaean) [inline opm/simulators/wells/WellInterface_impl.hpp:1420]: @atgeirr , do you have some comment regarding how to determine the group rate target is zero?

Maybe we should not compared directly with `0.0` to consider the rounding error. 

There is more code in the group rate target determination, maybe they can go in a separate function.

File: opm/simulators/wells/WellInterface_impl.hpp
Comment (steink) [reply opm/simulators/wells/WellInterface_impl.hpp:1416]: Thanks, good suggestion.

File: opm/simulators/wells/MultisegmentWell_impl.hpp
Comment (steink) [inline opm/simulators/wells/MultisegmentWell_impl.hpp:162]: I don't think setting the `WQTotal` of first segment explicitly to zero for zero-target wells is strictly necessary below, so have left out the checking for zero targets from groups here.

File: opm/simulators/wells/MultisegmentWell_impl.hpp
Comment (steink) [inline opm/simulators/wells/MultisegmentWell_impl.hpp:687]: Same comment as above.

File: opm/simulators/wells/StandardWell_impl.hpp
Comment (steink) [inline opm/simulators/wells/StandardWell_impl.hpp:697]: Same comment as for MS-wells

File: opm/simulators/wells/StandardWell_impl.hpp
Comment (steink) [inline opm/simulators/wells/StandardWell_impl.hpp:1189]: I think it's a bit strange that zero target wells should be solved with a much stricter tollerance than others. For convenience, I've also skipped checking for zero targets from groups here.

**Comment by steink** [comment]
I've restructered everything here, and renamed the original functions `wellUnderZeroRateTarget` and `stoppedOrZeroRateTarget` to `wellUnderZeroRateTargetIndividual` and `stoppedOrZeroRateTargetIndividual`, respectively. The functions `wellUnderZeroRateTarget` and `stoppedOrZeroRateTarget` now also include checking for zero targets from groups. I've updated the usage of these functions where I see it necessary (like forming of control equations and in the `iterateWellEqWithSwitching`-versions). 
Since the checking for zero targets from groups is a bit more involved, I've skipped this where I don't think it is important (see comments). Hence, these places now call the `*Individual` versions and the behaviour is unchanged from master.

**Comment by GitPaean** [comment]
jenkins build this please

File: opm/simulators/wells/MultisegmentWell_impl.hpp
Comment (GitPaean) [reply opm/simulators/wells/MultisegmentWell_impl.hpp:162]: It can be debatable. If we want to be a little more systematical, we probably want to remove this `stop_or_zero_rate_target` by all if it is not necessary. 

I think it is not a problem when it is with the correct sign, and it might be a problem when the sign of the `WQTotal` switches, which is easy for the wells without a zero rate target. 

So far we do not know how to handle wells with the opposite production/injection direction.

File: opm/simulators/wells/StandardWell_impl.hpp
Comment (GitPaean) [reply opm/simulators/wells/StandardWell_impl.hpp:1189]: There might be different approaches to address this, but it improves the results for some problematic cases. 

https://github.com/OPM/opm-simulators/pull/4572
https://github.com/OPM/opm-simulators/pull/4556

Possibly we should use stricter tolerance by all.

**Comment by GitPaean** [comment]
> I've restructered everything here, and renamed the original functions `wellUnderZeroRateTarget` and `stoppedOrZeroRateTarget` to `wellUnderZeroRateTargetIndividual` and `stoppedOrZeroRateTargetIndividual`, respectively. The functions `wellUnderZeroRateTarget` and `stoppedOrZeroRateTarget` now also include checking for zero targets from groups. I've updated the usage of these functions where I see it necessary (like forming of control equations and in the `iterateWellEqWithSwitching`-versions). Since the checking for zero targets from groups is a bit more involved, I've skipped this where I don't think it is important (see comments). Hence, these places now call the `*Individual` versions and the behaviour is unchanged from master.

Thanks for the update.  Will need to think through and get back with suggestions as a whole.

**Comment by GitPaean** [comment]
At the same time, can you check the Jenkins failures to see whether the failures are relevant?  Especially the ones with related to `MOD4_GRP`.

**Comment by steink** [comment]
> At the same time, can you check the Jenkins failures to see whether the failures are relevant? Especially the ones with related to `MOD4_GRP`.

Yes, will look into what's going on with these.

**Comment by steink** [comment]
Have looked at the four failing tests here. _NETWORK-01-REROUTE__ and _NETWORK-01-REROUTE_STD_ match very closely. For _MOD4_GRP_GEFAC_ there is slightly different time-stepping with the PR, but otherwise close match. 

Looked a bit more closely into _MOD4_GRP_. This is a very problematic test where both MASTER and PR struggles a lot towards the end with control switching oscillations resulting in extremely short time steps. PR struggles a bit more than MASTER here, but since both show very erratic behavour, I can'r really faviour one above the other. Besides this last problematic part, PR and MASTER match closely.

It's probably worthwhile to look more closely into why _MOD4_GRP_ is so problematic, and try to figure out whether the problem is with flow or with the test itself. Currently, it's not well suited as a test (due to the problems, it's also quite computationally expensive).

**Comment by GitPaean** [comment]
jenkins build this please

**Review by GitPaean — CHANGES_REQUESTED** [review]
Sorry for the delay.  I left some comments regarding how to organize the code.  Let me know if you need some help or want some discussion. 

I do realize this PR is a incremental change that fixes things without breaking other things, while still trying to make the code usage more systematical or reduce the number of the functions whenever possible. 

Let me know.

File: opm/simulators/wells/WellInterface_impl.hpp
Comment (steink) [reply opm/simulators/wells/WellInterface_impl.hpp:1425]: Thanks, will change to `target_scale` (since the output is a scaling factor).

File: opm/simulators/wells/WellInterface_impl.hpp
Comment (steink) [reply opm/simulators/wells/WellInterface_impl.hpp:1450]: This was was my first intention, but was not able to do it without a lot of other code-changes. If you see some smarter way to move it there, I'll be happy to.

File: opm/simulators/wells/StandardWell_impl.hpp
Comment (steink) [reply opm/simulators/wells/StandardWell_impl.hpp:1189]: Thanks, would be happy to discuss stricter well-tolerances overall.  Especially for standard wells I think this is a good idea!

File: opm/simulators/wells/MultisegmentWell_impl.hpp
Comment (steink) [reply opm/simulators/wells/MultisegmentWell_impl.hpp:162]: > I think it is not a problem when it is with the correct sign, and it might be a problem when the sign of the WQTotal switches, which is easy for the wells without a zero rate target.

I beleive `WQTotal` is not allowed to switch sign in the Newton update, so it either has the right sign or is zero.

File: opm/simulators/wells/WellInterface_impl.hpp
Comment (GitPaean) [reply opm/simulators/wells/WellInterface_impl.hpp:1425]: Then it is the function name `getGroupProductionTargetRate` needs to be updated. You can keep the `scale` variable name here.

**Comment by steink** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
jenkins build this please

**Review by GitPaean — APPROVED** [review]

**Review by atgeirr — APPROVED** [review]
Looks good and can be merged assuming the test failures have been checked.

For later: Will look at improving some interfaces to possibly reduce the need to pass `const Simulator&` around, and maybe avoid using logging and exceptions in the TargetCalculator.

**Comment by GitPaean** [comment]
The testing failures need to be double checked.

**Comment by GitPaean** [comment]
Not remembering what happened with the regression test before the refactoring, will create a PR for testing purpose.

**Comment by steink** [comment]
I've gone through the tests and think it looks OK. For the majority of the failures

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/steink/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/steink/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#467886" vlink="#96607D">



-- | -- | --
NETWORK-01_STANDARD | OK |  
NETWORK-01-REROUTE | OK |  
NETWORK-01-REROUTE_STD | OK |  
UDQ_WCONPROD | check | OK, different time-stepping
UDQ_ACTIONX | check | OK, different time-stepping
PYACTION_WCONPROD | check |  
9_3A_GINJ_REIN-G_MSW | tstep | OK, different time-stepping
9_3B_GINJ_GAS_EXPORT_MSW | tstep | OK, different time-stepping
9_3C_GINJ_GAS_GCONSUMP_MSW | check | OK, different time-stepping
9_3E_GAS_MIN_EXPORT_STW | OK |  
9_3D_GINJ_GAS_MAX_EXPORT_MSW | check | OK, different time-stepping
9_4A_WINJ_MAXWRATES_MAXBHP_GCONPROD_1L_STW | tstep | Quite different   results/time-stepping, time-steps much too long to catch dynamics accurately
9_4A_WINJ_MAXWRATES_MAXBHP_GCONPROD_1L_MSW | tstep | Quite different   results/time-stepping, time-steps much too long to catch dynamics accurately
9_4C_WINJ_GINJ_VREP-W_REIN-G_STW | check | OK, different time-stepping
9_4B_WINJ_VREP-W_MSW | tstep | OK, different time-stepping
9_4D_WINJ_GINJ_GAS_EXPORT_STW | tstep | OK, different time-stepping
9_4C_WINJ_GINJ_VREP-W_REIN-G_MSW | check | Different time-stepping, mostly   close match, except master violates group oil rate and pr violates group gas   rate
9_4D_WINJ_GINJ_GAS_EXPORT_MSW | check | Different time-stepping, mostly   close match, except master violates group oil rate
MOD4_UDQ_ACTIONX | tstep | OK,  slightly different time-stepping
ACTIONX_GCONPROD | rsm? | Not sure why this triggered   failure, results are almost identical
MOD4_GRP | tstep | OK, different   time-stepping,problematic case, but pr less problematic
MOD4_GRP_GEFAC | tstep | OK, different time-stepping
KRNUM-03X | check | OK, different time-stepping   (very long time-steps vs dynamics)
KRNUM-03Y | check | OK, different time-stepping   (very long time-steps vs dynamics)



</body>

</html>
[NETWORK-01](https://ci.opm-project.org/job/opm-simulators-PR-builder/6129/testReport/(root)/mpi/compareECLFiles_flow_NETWORK_01/)	OK	
NETWORK-01_STANDARD	OK	
NETWORK-01-REROUTE	OK	
NETWORK-01-REROUTE_STD	OK	
UDQ_WCONPROD	check	OK, different time-stepping
UDQ_ACTIONX	check	OK, different time-stepping
PYACTION_WCONPROD	check	
9_3A_GINJ_REIN-G_MSW	tstep	OK, different time-stepping
9_3B_GINJ_GAS_EXPORT_MSW	tstep	OK, different time-stepping
9_3C_GINJ_GAS_GCONSUMP_MSW	check	OK, different time-stepping
9_3E_GAS_MIN_EXPORT_STW	OK	
9_3D_GINJ_GAS_MAX_EXPORT_MSW	check	OK, different time-stepping
9_4A_WINJ_MAXWRATES_MAXBHP_GCONPROD_1L_STW	tstep	Quite different results/time-stepping, time-steps much too long to catch dynamics accurately
9_4A_WINJ_MAXWRATES_MAXBHP_GCONPROD_1L_MSW	tstep	Quite different results/time-stepping, time-steps much too long to catch dynamics accurately
9_4C_WINJ_GINJ_VREP-W_REIN-G_STW	check	OK, different time-stepping
9_4B_WINJ_VREP-W_MSW	tstep	OK, different time-stepping
9_4D_WINJ_GINJ_GAS_EXPORT_STW	tstep	OK, different time-stepping
9_4C_WINJ_GINJ_VREP-W_REIN-G_MSW	check	Different time-stepping, mostly close match, except master violates group oil rate and pr violates group gas rate
9_4D_WINJ_GINJ_GAS_EXPORT_MSW	check	Different time-stepping, mostly close match, except master violates group oil rate
MOD4_UDQ_ACTIONX	tstep	OK,  slightly different time-stepping
ACTIONX_GCONPROD	rsm?	Not sure why this triggered failure, results are almost identical
MOD4_GRP	tstep	OK, different time-stepping,problematic case, but pr less problematic
MOD4_GRP_GEFAC	tstep	OK, different time-stepping
KRNUM-03X	check	OK, different time-stepping (very long time-steps vs dynamics)
KRNUM-03Y	check	OK, different time-stepping (very long time-steps vs dynamics)

**Comment by steink** [comment]
I've gone through the failures and think they look OK. For most of them it's just a result of different time-stepping/different number of time-steps which in many cases leads to quite different results since time-steps are so large relative to the dynamics. 
Did not check `PYACTION_WCONPROD`.



































































































<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/steink/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/steink/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#467886" vlink="#96607D">


  |   |  
-- | -- | --
NETWORK-01 | OK |  
NETWORK-01_STANDARD | OK |  
NETWORK-01-REROUTE | OK |  
NETWORK-01-REROUTE_STD | OK |  
UDQ_WCONPROD | check | OK, different time-stepping
UDQ_ACTIONX | check | OK, different time-stepping
PYACTION_WCONPROD | check |  
9_3A_GINJ_REIN-G_MSW | tstep | OK, different time-stepping
9_3B_GINJ_GAS_EXPORT_MSW | tstep | OK, different time-stepping
9_3C_GINJ_GAS_GCONSUMP_MSW | check | OK, different time-stepping
9_3E_GAS_MIN_EXPORT_STW | OK |  
9_3D_GINJ_GAS_MAX_EXPORT_MSW | check | OK, different time-stepping
9_4A_WINJ_MAXWRATES_MAXBHP_GCONPROD_1L_STW | tstep | Quite different   results/time-stepping, time-steps much too long to catch dynamics accurately
9_4A_WINJ_MAXWRATES_MAXBHP_GCONPROD_1L_MSW | tstep | Quite different   results/time-stepping, time-steps much too long to catch dynamics accurately
9_4C_WINJ_GINJ_VREP-W_REIN-G_STW | check | OK, different time-stepping
9_4B_WINJ_VREP-W_MSW | tstep | OK, different time-stepping
9_4D_WINJ_GINJ_GAS_EXPORT_STW | tstep | OK, different time-stepping
9_4C_WINJ_GINJ_VREP-W_REIN-G_MSW | check | Different time-stepping, mostly   close match, except master violates group oil rate and pr violates group gas   rate
9_4D_WINJ_GINJ_GAS_EXPORT_MSW | check | Different time-stepping, mostly   close match, except master violates group oil rate
MOD4_UDQ_ACTIONX | tstep | OK,  slightly different time-stepping
ACTIONX_GCONPROD | rsm? | Not sure why this triggered   failure, results are almost identical
MOD4_GRP | tstep | OK, different   time-stepping,problematic case, but pr less problematic
MOD4_GRP_GEFAC | tstep | OK, different time-stepping
KRNUM-03X | check | OK, different time-stepping   (very long time-steps vs dynamics)
KRNUM-03Y | check | OK, different time-stepping   (very long time-steps vs dynamics)



</body>

**Comment by GitPaean** [comment]
The following is some notes for information. 


The thing is a little worrying is that, when we check the group target for a well, the group target might not be ready. 

I looked at the case KRNUM-03X, 

the SCHUEDULE setup for the well `F-1H` is as follows, with the current code, at the beginning, it will think F-1H is under zero injection group target, probably because the production rate is not ready yet so VREP get 0. 

```
WELSPECS
 'F-1H'  'S1'   3    3      1*   GAS     1*   1*   SHUT 1* 1* 1* /
/

GCONINJE
 'S1'   'GAS'	 'VREP'  3*	 1.020    'NO'  5* /
/

WCONINJE
-- Well_name    Type    Status  Ctrl    SRate1  Rrate   BHP     THP     VFP
  'F-1H'        GAS   OPEN    GRUP    8.0E6    1*      300.0	1*	1*     /
/
```

The VREP group control with above similar setup should be causes for many regression failures.  Because ideally, for many cases, there should be no wells under zero rate target at all and should no be affected.

**Comment by GitPaean** [comment]
> Yes, but don't you think it's better in this case to treat it properly as a zero rate well than trying to solve it with an individual zero rate (as in master) which even might give singular system?

It took some thinking to understand your point. Basically, you meant that without the checking, we will solve it with zero rate target in the master branch.  I think your point is valid and I agree. 

My point was that `stopppedOrZeroRateTarget()` can be improved to reflect better. Or we should have different ways in handling the VREP injections, but that should be outside of the scope of this PR.

**Comment by steink** [comment]
>It took some thinking to understand your point. Basically, you meant that without the checking, we will solve it with zero rate target in the master branch. I think your point is valid and I agree.

Sorry, was trying to finnish the sentence before a meeting ;)

>My point was that stopppedOrZeroRateTarget() can be improved to reflect better. Or we should have different ways in handling the VREP injections, but that should be outside of the scope of this PR.

I see you point, I agree.

**Comment by GitPaean** [comment]
For the test failure, `ACTIONX_GCONPROD`, the message is not very clear. 

```
Loading rft file /var/lib/jenkins/workspace/opm-simulators-PR-builder/deps/opm-tests/actionx/opm-simulation-reference/flow/ACTIONX_GCONPROD.RFT  .... done
Loading rft file /var/lib/jenkins/workspace/opm-simulators-PR-builder/mpi/build-opm-simulators/tests/results/flow+actionx_gconprod/ACTIONX_GCONPROD.RFT  .... done

Well: OP01 date: 2017/12/31
Comparing: TIME ...  done.
Comparing: DATE ...  done.
Comparing: WELLETC ...  done.
Comparing: CONIPOS ...  done.
Comparing: CONJPOS ...  done.
Comparing: CONKPOS ...  done.
Comparing: HOSTGRID ...  done.
Comparing: DEPTH ...  done.
Comparing: PRESSURE ...  done.
Comparing: SWAT ...  done.
Comparing: SGAS ...  done.

Program threw an exception: [/var/lib/jenkins/workspace/opm-simulators-PR-builder/deps/opm-common/test_util/compareECL.cpp:265] 1 errors encountered in comparisons.
```

**Comment by GitPaean** [comment]
With some discussion and help from @bska , the regression failure for `ACTIONX_GCONPROD` is related to RUNSUM keyword.  

Since the simulation results look generally fine and for this failure, we can update reference.  But we need to wait for the release processing to unfreeze the reference result update.

**Comment by GitPaean** [comment]
@steink , can you update the PR to fix the merging conflicts, then we can check how to adjust and update the reference so that we can get the PR in.

**Comment by steink** [comment]
> @steink , can you update the PR to fix the merging conflicts, then we can check how to adjust and update the reference so that we can get the PR in.
Thought I was able to fix conflicts here, but rebase attempt ended up in big mess. Maybe you can have a look @GitPaean?

**Comment by akva2** [comment]
please grab rebased branch from https://github.com/akva2/opm-simulators/tree/check_zero_target_from_group. note I squashed the previously botched rebasing, makes no sense to keep that history.

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by steink** [comment]
> please grab rebased branch from https://github.com/akva2/opm-simulators/tree/check_zero_target_from_group. note I squashed the previously botched rebasing, makes no sense to keep that history.

Thanks @akva2 !

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
Some how the regression related to `PYACTION_WCONPROD` does not look severe anymore. 

https://ci.opm-project.org/job/opm-simulators-PR-builder/6269/testReport/junit/(root)/mpi/compareECLFiles_flow_PYACTION_WCONPROD/

Then most likely, we do not need to adjust to make the regression test to work.

**Comment by GitPaean** [comment]
Will update_reference for investigation purpose.  If things look okay, will get prepared to get this PR in.

**Comment by GitPaean** [comment]
jenkins build this update_data please

**Comment by GitPaean** [comment]
> Some how the regression related to `PYACTION_WCONPROD` does not look severe anymore.
> 
> https://ci.opm-project.org/job/opm-simulators-PR-builder/6269/testReport/junit/(root)/mpi/compareECLFiles_flow_PYACTION_WCONPROD/
> 
> Then most likely, we do not need to adjust to make the regression test to work.

I took this back. It still remains a problem with more investigation.

**Comment by GitPaean** [comment]
jenkins build this update_data please

**Comment by bska** [comment]
jenkins build this opm-tests=1168 please

**Comment by bska** [comment]
The new reference solutions have been installed on the CI system.  Merging this now to activate the new logic.  Thanks a lot for seeing this through!

**Pull request merged**
